lib/HappyPlugin.js

```diff
@@ -66,9 +66,10 @@ function HappyPlugin(userConfig) {
 }
 
 HappyPlugin.prototype.apply = function(compiler) {
-  var that, engageWatchMode;
+  var that, webpack4PluginSupport, engageWatchMode, watchRunCallback, doneCallback;
 
   that = this;
+  webpack4PluginSupport = 'hooks' in compiler;
```

Comment:
I'd rather have an abstraction that decides whether to call the `plugin` function directly or to use the `hooks` API instead.

A sample implementation would be:

```javascript
function tapInto(target, hook, f) {
  if (webpack4) {
    const webpack4Hook = resolveHook(hook) // e.g. watch-run -> watchRun
    target.hooks[webpack4Hook].tap('HappyPlugin', f)
  } else {
    target.plugin(hook, f)
}

tapInto(compiler, 'watch-run', ...)
tapInto(compiler, 'compilation', compilation => {
  tapInto(compilation, 'failed-module', ...)
})
```

That would help keep the code remain readable.